pkg/kubectl/resource/visitor.go

```diff
@@ -0,0 +1,422 @@
+}
+
+// Info contains temporary info to execute a REST call, or show the results
+// of an already completed REST call.
+type Info struct {
+	Client    RESTClient
+	Mapping   *meta.RESTMapping
+	Namespace string
+	Name      string
+
+	// Optional, this is the most recent value returned by the server if available
+	runtime.Object
+	// Optional, this is the most recent resource version the server knows about for
+	// this type of resource. It may not match the resource version of the object,
+	// but if set it should be equal to or newer than the resource version of the
+	// object (however the server defines resource version).
+	ResourceVersion string
+}
+
+// NewInfo returns a new info object
+func NewInfo(client RESTClient, mapping *meta.RESTMapping, namespace, name string) *Info {
+	return &Info{
+		Client:    client,
+		Mapping:   mapping,
+		Namespace: namespace,
+		Name:      name,
+	}
+}
+
+// Visit implements Visitor
+func (i *Info) Visit(fn VisitorFunc) error {
+	return fn(i)
+}
+
+func (i *Info) Get() error {
+	obj, err := NewHelper(i.Client, i.Mapping).Get(i.Namespace, i.Name)
+	if err != nil {
+		return err
+	}
+	i.Object = obj
+	i.ResourceVersion, _ = i.Mapping.MetadataAccessor.ResourceVersion(obj)
+	return nil
+}
+
+// Watch returns server changes to this object after it was retrieved.
+func (i *Info) Watch(resourceVersion string) (watch.Interface, error) {
+	return NewHelper(i.Client, i.Mapping).WatchSingle(i.Namespace, i.Name, resourceVersion)
+}
+
+// ResourceMapping returns the mapping for this resource and implements ResourceMapping
+func (i *Info) ResourceMapping() *meta.RESTMapping {
+	return i.Mapping
+}
+
+// VisitorList implements Visit for the sub visitors it contains. The first error
+// returned from a child Visitor will terminate iteration.
+type VisitorList []Visitor
+
+// Visit implements Visitor
+func (l VisitorList) Visit(fn VisitorFunc) error {
+	for i := range l {
+		if err := l[i].Visit(fn); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// EagerVisitorList implements Visit for the sub visitors it contains. All errors
+// will be captured and returned at the end of iteration.
+type EagerVisitorList []Visitor
+
+// Visit implements Visitor, and gathers errors that occur during processing until
+// all sub visitors have been visited.
+func (l EagerVisitorList) Visit(fn VisitorFunc) error {
+	errs := []error(nil)
+	for i := range l {
+		if err := l[i].Visit(func(info *Info) error {
+			if err := fn(info); err != nil {
+				errs = append(errs, err)
+			}
+			return nil
+		}); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return util.SliceToError(errs)
+}
+
+// PathCollector visits a given path and returns an object representing the file
+// at that path.
+type PathCollector struct {
+	*Mapper
+	// The file path to load
+	Path string
+	// Whether to ignore files that are not recognized as API objects
+	IgnoreErrors bool
+}
+
+func (v *PathCollector) Visit(fn VisitorFunc) error {
```

Comment:
Having them in one pass tends to make some of the other stuff simpler (if only because Result is acting as a Collector).  I can leave as Visitor now and we can assess later.

----- Original Message -----
> > +		}
> > +	}
> > +	return util.SliceToError(errs)
> > +}
> > +
> > +// PathCollector visits a given path and returns an object representing
> > the file
> > +// at that path.
> > +type PathCollector struct {
> > +	*Mapper
> > +	// The file path to load
> > +	Path string
> > +	// Whether to ignore files that are not recognized as API objects
> > +	IgnoreErrors bool
> > +}
> > +
> > +func (v *PathCollector) Visit(fn VisitorFunc) error {
> 
> Thanks for the rename, but I don't know if the rename makes while this is
> still doing collection and visitation in a single pass. Also, I don't think
> the PathCollector and DirectoryCollector should use a different convention
> than URLVisitor and StreamVisitor. So, I think you should either revert the
> rename and leave them as they were (i.e., *Visitor), or change all similar
> receivers, and split the generation of the list of Infos and the visitation
> of them into 2 passes.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3152/files#r22688434